baremetal_network_provisioning/ml2/extensions/switch.py

```diff
@@ -12,29 +12,225 @@
+        for key, value in access_parameters.iteritems():
+            body[key] = value
+        switch = self._update_dict(body, switch_dict)
+        snmp_client = discovery_driver.SNMPDiscoveryDriver(switch_dict)
+        bnp_switch = snmp_client.discover_switch()
+        if bnp_switch.get('mac_addr'):
+            switch['mac_address'] = bnp_switch.get('mac_addr')
+            switch['status'] = 'ENABLED'
+        db_switch = db.add_bnp_phys_switch(self._dbcontext, switch)
+        if bnp_switch.get('ports'):
+            self._add_physical_port(db_switch.get('id'),
+                                    bnp_switch.get('ports'))
+        return switch
+
+    def _add_physical_port(self, switch_id, ports):
+        for port in ports:
+            port['switch_id'] = switch_id
+            status = const.Port_status.get(port['port_status'])
+            port['port_status'] = status
+            db.add_bnp_phys_switch_port(self._dbcontext, port)
+
+    def _validate_snmp_parameters(self, access_parameters):
+        if not access_parameters.get('write_community'):
+            raise exc.BadRequest(
+                resource="switch",
+                reason="'write_community' not found in request body")
+
+    def _validate_snmpv3_parameters(self, access_parameters):
+        if not access_parameters.get('security_name'):
+            raise exc.BadRequest(
+                resource="switch",
+                reason="'security_name' not found in request body")
+        if access_parameters.get('auth_protocol'):
+            if access_parameters.get('auth_protocol').lower(
+            ) not in const.SUPPORTED_AUTH_PROTOCOLS:
+                raise exc.BadRequest(
+                    resource="switch",
+                    reason="Auth Protocol '%s' is not supported" %
+                    access_parameters['auth_protocol'])
+            elif not access_parameters.get('auth_key'):
+                raise exc.BadRequest(
+                    resource="switch",
+                    reason="Auth Key is required for Auth Protocol %s" %
+                    access_parameters['auth_protocol'])
+        if access_parameters.get('priv_protocol'):
+            if access_parameters.get('priv_protocol').lower(
+            ) not in const.SUPPORTED_PRIV_PROTOCOLS:
+                raise exc.BadRequest(
+                    resource="switch",
+                    reason="Priv Protocol '%s' is not supported" %
+                    access_parameters['priv_protocol'])
+            elif not access_parameters.get('priv_key'):
+                raise exc.BadRequest(
+                    resource="switch",
+                    reason="Priv Key is required for Priv Protocol %s" %
+                    access_parameters['priv_protocol'])
+
+    def update(self, request, id):
+        try:
+            body = request.json_body
+        except json_scanner.JSONDecodeError:
+            raise exc.BadRequest(
+                resource="switch",
```

Comment:
make this block as a single function as it is repeating many times raise exc.BadRequest(
+                resource="switch",
+                reason="invalid JSON body")